app/models/category.rb

```diff
@@ -1172,6 +1173,44 @@ def clear_site_cache
   def on_custom_fields_change
     clear_site_cache
   end
+
+  def self.reorder_categories!
+    categories =
+      if SiteSetting.fixed_category_positions
+        Category.order(:position)
+      else
+        Category
+          .left_outer_joins(:featured_topics)
+          .group("categories.id")
+          .order("max(topics.bumped_at) DESC NULLS LAST")
+      end.order(:id).pluck(:id, :parent_category_id)
+
+    # Start with an empty "root" category
+    ordered_categories = [[nil, nil]]
+
+    # Insert all subcategories after the parent category
+    ordered_categories.each_with_index do |(id, _), index|
+      if subcategory_ids = categories.filter { |_, parent_id| parent_id == id }.presence
+        ordered_categories.insert(index + 1, *subcategory_ids)
+      end
+    end
+
+    # Drop the "root" category
+    ordered_categories.shift
```

Comment:
I didn't know that you could iterate over an array while doing updates. I know that doesn't work with a `Hash`. TIL.

This is quite slow though. `filter` and `insert` are both `O(n)` and that's happening for each category. You're also not checking for loops. Here's an alternative:

```suggestion
    categories_by_parent = categories.group_by(&:second)

    ordered_categories = []
    traversed_set = Set.new
    to_traverse = categories_by_parent[nil]

    until to_traverse.empty?
      x = to_traverse.pop

      unless traversed_set.include?(x)
        ordered_categories << x
        traversed_set.add(x)
        to_traverse.concat((categories_by_parent[x.first] || []).reverse)
      end
    end
```

Reply:
Yes, I realized that I could do a better job by "precomputing" `subcategory_ids`. Also, copying the array everytime by using `insert` is not so efficient either. However, it was super easy and quick to implement which helped me with my experimentation.

In my 10k categories test it was fast enough even if it is O(N^2).